src/rgw/driver/rados/rgw_notify.cc

```diff
@@ -1006,6 +1004,33 @@ static inline bool notification_match(reservation_t& res,
         ldpp_dout(res.dpp, 1) << "ERROR: failed to parse reservation id. error: " << ret << dendl;
         return ret;
       }
+    } else {
+      // no reservation - try to connect to endpoint instead
+      try {
+        const auto endpoint = RGWPubSubEndpoint::create(
+	  topic_cfg.dest.push_endpoint,
+	  topic_cfg.dest.arn_topic,
+	  RGWHTTPArgs(topic_cfg.dest.push_endpoint_args, dpp),
+	  dpp->get_cct());
+        ldpp_dout(res.dpp, 20) << "INFO: push endpoint created: "
+			       << topic_cfg.dest.push_endpoint << dendl;
+        const auto ret = endpoint->ping(dpp->get_cct(), res.yield);
```

Comment:
With this ping, if there is a transient `short lived` error at same time, we will be failing the request and not allowing the kafka retry which happens otherwise
ideally there should be a way once we call `kafka_produce` after timeout to verify if the broker was down and then return the error?

Reply:
note that this is only for syncronouse notifications, so there is no retry mechanism at our level of the code. if you want a to retry, usepersistent notifications.
if there is an error in "produce()" or "poll()", this is after the kafka library deicded that there is an error (e.g. exhausted internal retries).
since wew mimic here the 2 phases, we cannot try to send the notification, because, if we succeed, and then the S3 op fails, there is no way to rollback the notification.

Reply:
i still do not understand how is this solving the issue,
you call `ping` which checks the `conn->status` in `publish_reserve()`.
the `conn->status` gets updated when kafka is `idle` indicating the broker is down (it also gets updated for other transient issues, but i guess those are irrelevant). As soon as it gets `idle` we also delete the connection.
so now the next s3 requests will not find the connections and again do a new connection when `RGWPubSubKafkaEndpoint::Create` is called and new connection will be emplaced in map and`status` for new connection is set to `ok`, so the new connection will also wait for `idle_timeout` and until then the `conn->status` will `ok`, but since broker is down the new s3 request will wait until timeout before returning,  so the same issue will still exist ??

Reply:
as long as there are notifications to be sent to a broker, it is not considered to be idle (even if it is "down").
we set `conn->timestamp` inside the `publish_internal()` call regardless of connection state.

the "idle" mechanism is meant to eb a garbage collector for the case that we delete all topics and notifications that used to send to this broker, but the kafka connection is still there.

Reply:
> as long as there are notifications to be sent to a broker, it is not considered to be idle (even if it is "down"). we set `conn->timestamp` inside the `publish_internal()` call regardless of connection state.
> 
> the "idle" mechanism is meant to eb a garbage collector for the case that we delete all topics and notifications that used to send to this broker, but the kafka connection is still there.

as per code here (https://github.com/ceph/ceph/blob/main/src/rgw/rgw_kafka.cc#L467-L471), if there is timeout for any one of the sent event `connections.erase` which will invoke the `destroy` and remove the connection from list. So again all the subsequent connect will go through same timeout logic ?

Reply:
> if there is timeout for any one of the sent event

the code above checks idleness for the connection itself, not a timeout for an event
and a connection is considered idle only if there aren't any events that we ***try*** to send to it for more than couple of minutes.
see: https://github.com/ceph/ceph/blob/main/src/rgw/rgw_kafka.cc#L354